# CMOR 350 HW 7 Question 4 Part A
The goal is to find the occupancy times of this rate matrix, in order to calculate the amount of time Machine 1 is blocked in the first hour of operation.

Since the engines operate for 200 hrs on average, and they are exponentially distributed and  $\lambda=\frac{1}{200}$

Here's the rate matrix.

In [15]:
l = 1/200

r_matrix = [
    #1,     2,       3,     4,      5,      6,      7,      8       9
    [0,     2*l,     2*l,   0,      0,      0,      0,      0,      0],     #1
    [0,     0,       0,     l,      2*l,    0,      0,      0,      0],     #2
    [0,     0,       0,     0,      2*l,    l,      0,      0,      0],     #3 
    [0,     0,       0,     0,      0,      0,      2*l,    0,      0],     #4
    [0,     0,       0,     0,      0,      0,      l,      l,      0],     #5 
    [0,     0,       0,     0,      0,      0,      0,      2*l,    0],     #6 
    [0,     0,       0,     0,      0,      0,      0,      0,      l],     #7 
    [0,     0,       0,     0,      0,      0,      0,      0,      l],     #8 
    [0,     0,       0,     0,      0,      0,      0,      0,      0],     #9

]
r_matrix

[[0, 0.01, 0.01, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0.005, 0.01, 0, 0, 0, 0],
 [0, 0, 0, 0, 0.01, 0.005, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0.01, 0, 0],
 [0, 0, 0, 0, 0, 0, 0.005, 0.005, 0],
 [0, 0, 0, 0, 0, 0, 0, 0.01, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0.005],
 [0, 0, 0, 0, 0, 0, 0, 0, 0.005],
 [0, 0, 0, 0, 0, 0, 0, 0, 0]]

In [16]:
for row in r_matrix:
    print(sum(row))

0.02
0.015
0.015
0.01
0.01
0.01
0.005
0.005
0


In [17]:
row_sum = []
for row in r_matrix:
    row_sum.append(sum(row))
nu = max(row_sum)
row_sum, nu

([0.02, 0.015, 0.015, 0.01, 0.01, 0.01, 0.005, 0.005, 0], 0.02)

In [18]:
p_hat = []
for row in r_matrix:
    p_hat.append([x / nu for x in row])
p_hat

[[0.0, 0.5, 0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.25, 0.5, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.5, 0.25, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.25, 0.25, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.25],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.25],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]

In [19]:
for row in p_hat:
    print(sum(row))

1.0
0.75
0.75
0.5
0.5
0.5
0.25
0.25
0.0


In [20]:
for i in range(len(p_hat)):
    p_hat[i][i] = 1 - sum(p_hat[i])
p_hat

[[0.0, 0.5, 0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.25, 0.0, 0.25, 0.5, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.25, 0.0, 0.5, 0.25, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.5, 0.0, 0.0, 0.5, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0.25, 0.25, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0.5, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.75, 0.0, 0.25],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.75, 0.25],
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]]

In [21]:
for row in p_hat:
    print(sum(row))

1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0


### $\hat P$ Is Now Created

Now we want 
$$\begin{align}
M(T)&=\sum_{k=0}^\infin\frac{1}{\nu}P(Y>k)\hat{P}^k\\
&=\sum_{k=0}^\infin\frac{1}{\nu}(1-\sum_{l=0}^ke^{-\nu T}\frac{(\nu T)^l}{l!}))\hat{P}^k
\end{align}$$
To get the Occupancy matrix for the first hour of operations.

Note - $\infin$ is too large, so we can approximate it with 
$$max(\nu t + 5\sqrt{\nu t},20)$$

$$M(T) = \sum_{k=0}^{max(\nu t + 5\sqrt{\nu t},20)}\frac{1}{\nu}(1-\sum_{l=0}^ke^{-\nu T}\frac{(\nu T)^l}{l!}))\hat{P}^k$$


### Note; units
Our $\lambda$ are given in terms of hours, but we are asked about the first 6 hours. Thus, we'll be sure to look at `M(6)`



In [25]:
import numpy as np

p_hat=np.array(p_hat)

p_hat

array([[0.  , 0.5 , 0.5 , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ],
       [0.  , 0.25, 0.  , 0.25, 0.5 , 0.  , 0.  , 0.  , 0.  ],
       [0.  , 0.  , 0.25, 0.  , 0.5 , 0.25, 0.  , 0.  , 0.  ],
       [0.  , 0.  , 0.  , 0.5 , 0.  , 0.  , 0.5 , 0.  , 0.  ],
       [0.  , 0.  , 0.  , 0.  , 0.5 , 0.  , 0.25, 0.25, 0.  ],
       [0.  , 0.  , 0.  , 0.  , 0.  , 0.5 , 0.  , 0.5 , 0.  ],
       [0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.75, 0.  , 0.25],
       [0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.75, 0.25],
       [0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 1.  ]])

In [26]:
import math
import numpy as np

def calculate_M(T, nu, P_hat):
    max_k = round(max(nu*T + 5*math.sqrt(nu*T),20))
    print(max_k)
    total = 0
    for k in range(max_k):
        inner_sum = sum(math.exp(-nu * T + l * math.log(nu * T) - math.lgamma(l + 1)) for l in range(k + 1))
        total += (1 / nu) * (1 - inner_sum) * np.linalg.matrix_power(P_hat, k)
    return total

# Distribution of the first 60 min
occupancy_matrix_60min = calculate_M(6, nu, p_hat)
occupancy_matrix_60min

20


array([[5.65397816e+00, 1.67885636e-01, 1.67885636e-01, 1.68284188e-03,
        6.73136752e-03, 1.68284188e-03, 7.62943704e-05, 7.62943704e-05,
        9.24727016e-07],
       [0.00000000e+00, 5.73792098e+00, 0.00000000e+00, 8.56256597e-02,
        1.71251319e-01, 0.00000000e+00, 3.44197813e-03, 1.72098907e-03,
        3.90719122e-05],
       [0.00000000e+00, 0.00000000e+00, 5.73792098e+00, 0.00000000e+00,
        1.71251319e-01, 8.56256597e-02, 1.72098907e-03, 3.44197813e-03,
        3.90719122e-05],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 5.82354664e+00,
        0.00000000e+00, 0.00000000e+00, 1.74693297e-01, 0.00000000e+00,
        1.76006098e-03],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        5.82354664e+00, 0.00000000e+00, 8.73466487e-02, 8.73466487e-02,
        1.76006098e-03],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 5.82354664e+00, 0.00000000e+00, 1.74693297e-01,
        1.7

### E[Amount of Time plane will crash]
State's 4,6,7,8,9 cause the plane to crash

Thus, we are interested in the quanitity $m_{14}(6)+m_{15}(6)+m_{16}(6)+m_{17}(6)+m_{18}(6)+m_{19}(6)$

In [29]:
# The sum is the following
crash_time=occupancy_matrix_60min[1-1][4-1]
+occupancy_matrix_60min[1-1][5-1]
+occupancy_matrix_60min[1-1][6-1]
+occupancy_matrix_60min[1-1][7-1]
+occupancy_matrix_60min[1-1][8-1]
+occupancy_matrix_60min[1-1][9-1]

9.247270163507578e-07

Thus, in a 6 hour interval, $m_{14}(6)+m_{15}(6)+m_{16}(6)+m_{17}(6)+m_{18}(6)+m_{19}(6)\approx 9.25\times10^{-7}$ hours the plane is crashing


In [30]:
crash_time*60

0.1009705128607387

That's an expectation of about 0.1 minutes on average

In [31]:
crash_time/6

0.00028047364683538527

The likelihood that the plane crashes is $\approx 0.00028= 0.028\%$

In [32]:
1-crash_time/6

0.9997195263531646

The likelihood the plane does not crash is $\approx 0.9997$